src/Stratis.Bitcoin.Features.SmartContracts/Rules/SmartContractCoinViewRuleLogic.cs

```diff
@@ -82,6 +85,24 @@ public async Task RunAsync(Func<RuleContext, Task> baseRunAsync, RuleContext con
 
             this.cachedResults = this.executionCache.GetExecutionResult(block.GetHash());
 
+            if (this.cachedResults == null && context.ValidationContext.ChainedHeaderToValidate.Height <= this.lastCheckPointHeight)
```

Comment:
I'm not entirely sure I see how this check aligns with the objective in the title of the PR - unless this is only meant to operate on rewinds occurring below the last checkpoint height? We would expect those to be somewhat uncommon though, unless there are a number of reachable peers with differing chains, but those are then supposed to fail the checkpoint hash elsewhere in any case

Reply:
This relates to command-line initiated rewinds (`rewind=nnn`) and IBD - whenever we are re-synching blocks below the last checkpoint. On first IBD no contract execution data will exist, but after a rewind there will be pre-existing contract data that can be leveraged. Also, if we were to distribute the contract data with the release this same code will leverage that data instead of re-executing contracts.